### PR TITLE
Add validation

### PR DESCRIPTION
**Pull request by MateuszZacharecki**

### CODE REVIEW

**Comment by DawidPludowski** [comment]
@MateuszZacharecki pls remove `.idea` folder from the PR

File: hellsemble/hellsemble.py [hunk 1]
```diff
@@ -175,10 +176,12 @@ def __fit_estimators_sequential(
                 list of fitted estimators and the list of masks indicating
                 which observations were used during fit of the estimators.
         """
+        X_train, X_val, y_train, y_val = train_test_split(X, y, test_size = 0.25)
```
Comment (DawidPludowski) [inline hellsemble/hellsemble.py:179]: We should introduce `seed` parameter to the `fit` method and pass it to every `train_test_split` to ensure reproducibility

File: hellsemble/hellsemble.py [hunk 2]
```diff
@@ -210,6 +213,13 @@ def __fit_estimators_sequential(
                 X_fit[failed_observations_mask],
                 y_fit[failed_observations_mask],
             )
+
+            # Validate the ensemble
+            val_score = self.evaluate_hellsemble(X_val, y_val)
+            print(f'Validation score for sequential emsemble: {val_score}')
```
Comment (DawidPludowski) [inline hellsemble/hellsemble.py:219]: big "no" for `print` in classes; if it's important, we can:
* store training history in `self.meta` object or
* return it as an additional element

File: hellsemble/hellsemble.py [hunk 3]
```diff
@@ -210,6 +213,13 @@ def __fit_estimators_sequential(
...
+            if val_score >= 0.95:
```
Comment (DawidPludowski) [inline hellsemble/hellsemble.py:220]: it would be best if we make tihs `0.95` as an optional parameter in `fit`

File: hellsemble/hellsemble.py [hunk 4]
```diff
@@ -230,10 +240,12 @@ def __fit_estimators_greedy(
             list[np.ndarray]: list of masks indicating
                 which observations were used during fit of the estimators
         """
+        X_train, X_val, y_train, y_val = train_test_split(X, y, test_size = 0.25)
```
Comment (DawidPludowski) [inline hellsemble/hellsemble.py:243]: same as above

File: hellsemble/hellsemble.py [hunk 5]
```diff
@@ -284,14 +296,16 @@ def __fit_estimators_greedy(
...
-                fitting_history_entry = np.full((X.shape[0]), False)
+                fitting_history_entry = np.full((X_train.shape[0]), False)
                 fitting_history_entry[failed_observations_idx] = True
                 fitting_history.append(fitting_history_entry)
                 X_fit, y_fit = (
                     X_fit[failed_observations_mask],
                     y_fit[failed_observations_mask],
                 )
 
+                print(f'Validation score for greedy ensemble: {best_score}')
```
Comment (DawidPludowski) [inline hellsemble/hellsemble.py:307]: same as above

Comment (DawidPludowski) [reply hellsemble/hellsemble.py:307]: break on validation thresold is missing (?)

**Comment by DawidPludowski** [comment]
If I understand correctly, the logic for now is as follows:

* make train-val split before the start of the training 
* model in greedy approach **is still chosen based on train set only**
* threshold is taken into consideration for early-stopping

Honestly, I do not remember the actual goal of this [issue](https://github.com/KN-Modelowania-Matematycznego/hellsemble/issues/10) but I thought that the validation was made to improve the selection of the models in the greedy approach (to choose not overfitted one). Afaik this is not done in the current implementation. @jakub-piw, @azoz01, what are your opinions on that?

**Review by DawidPludowski — CHANGES_REQUESTED** [review]

File: hellsemble/hellsemble.py (hunk 2 above)
Comment (azoz01) [reply hellsemble/hellsemble.py:219]: The print should be removed. But some logger in DEBUG level should be ok

File: hellsemble/hellsemble.py (hunk 3 above)
Comment (azoz01) [reply hellsemble/hellsemble.py:220]: For the future: one fixed threshold for the metric across different datasets is quite risky. Depending on the metric and prior on y 0.95 can mean both good model and bad model

**Comment by azoz01** [comment]
> * model in greedy approach **is still chosen based on train set only**

I think that it should be chosen based on validation metrics. The choice based on the train can lead to the promotion of an overfitted model.

**Comment by MateuszZacharecki** [comment]
> * model in greedy approach **is still chosen based on train set only**

Models are chosen based on evaluations on valid sets and received scores based on that. Unless I can't see sth.

**Comment by DawidPludowski** [comment]
> > * model in greedy approach **is still chosen based on train set only**
> 
> Models are chosen based on evaluations on valid sets and received scores based on that. Unless I can't see sth.

My bad, I missed that `append`/`pop` logic in the inner loop

**Comment by DawidPludowski** [comment]
I added some fixes as in the commit message. 

Btw, the abstraction in `estimator_egnerator.py` is a bit incorrect (`reset_generator` is not a part of the abstraction, not sure if it's should @jakub-piw )

**Review by DawidPludowski — APPROVED** [review]
